sinker/generate_csv_sinker.go

```diff
@@ -0,0 +1,311 @@
+
+	loader *db.Loader
+	logger *zap.Logger
+	tracer logging.Tracer
+
+	stats *Stats
+}
+
+func NewGenerateCSVSinker(
+	sink *sink.Sinker,
+	destFolder string,
+	workingDir string,
+	bundleSize uint64,
+	bufferSize uint64,
+	loader *db.Loader,
+	logger *zap.Logger,
+	tracer logging.Tracer,
+) (*GenerateCSVSinker, error) {
+	blockRange := sink.BlockRange()
+	if blockRange == nil || blockRange.EndBlock() == nil {
+		return nil, fmt.Errorf("sink must have a stop block defined")
+	}
+
+	// create cursor in working Dir
+	stateStorePath := filepath.Join(workingDir, "state.yaml")
+	stateFileDirectory := filepath.Dir(stateStorePath)
+	if err := os.MkdirAll(stateFileDirectory, os.ModePerm); err != nil {
+		return nil, fmt.Errorf("create state file directories: %w", err)
+	}
+
+	stateDStore, err := dstore.NewStore(destFolder, "", "", false)
+	if err != nil {
+		return nil, err
+	}
+	stateStore, err := state.NewFileStateStore(stateStorePath, stateDStore, logger)
+	if err != nil {
+		return nil, fmt.Errorf("new file state store: %w", err)
+	}
+
+	s := &GenerateCSVSinker{
+		Shutter: shutter.New(),
+		Sinker:  sink,
+
+		fileBundlers: make(map[string]*bundler.Bundler),
+		stopBlock:    *blockRange.EndBlock(),
+
+		loader: loader,
+		logger: logger,
+		tracer: tracer,
+
+		stateStore: stateStore,
+		bundleSize: bundleSize,
+
+		stats: NewStats(logger),
+	}
+
+	csvOutputStore, err := dstore.NewStore(destFolder, "csv", "", false)
+	if err != nil {
+		return nil, err
+	}
+	tables := s.loader.GetAvailableTablesInSchemaList()
+
+	for _, table := range tables {
+		columns := s.loader.GetColumnsForTable(table)
+		fb, err := getBundler(table, s.Sinker.BlockRange().StartBlock(), s.stopBlock, bundleSize, bufferSize, csvOutputStore, workingDir, logger, columns)
+		if err != nil {
+			return nil, err
+		}
+		s.fileBundlers[table] = fb
+	}
+
+	return s, nil
+}
+
+func (s *GenerateCSVSinker) Run(ctx context.Context) {
+	s.stateStore.Start(ctx)
+	cursor, err := s.stateStore.ReadCursor(ctx)
+	if err != nil && !errors.Is(err, db.ErrCursorNotFound) {
+		s.Shutdown(fmt.Errorf("unable to retrieve cursor: %w", err))
+		return
+	}
+
+	s.Sinker.OnTerminating(s.Shutdown)
+	s.OnTerminating(func(err error) {
+		s.Sinker.Shutdown(err)
+		s.stats.LogNow()
+		s.logger.Info("csv sinker terminating", zap.Stringer("last_block_written", s.stats.lastBlock))
+		s.stats.Close()
+		s.stateStore.Shutdown(err)
+		s.CloseAllFileBundlers(err)
+	})
+
+	s.stats.OnTerminated(func(err error) { s.Shutdown(err) })
+
+	logEach := 15 * time.Second
+	if s.logger.Core().Enabled(zap.DebugLevel) {
+		logEach = 5 * time.Second
+	}
+
+	s.stats.Start(logEach, cursor)
+
+	s.logger.Info("starting postgres sink",
+		zap.Duration("stats_refresh_each", logEach),
+		zap.Stringer("restarting_at", cursor.Block()),
+		zap.String("database", s.loader.GetDatabase()),
+		zap.String("schema", s.loader.GetSchema()),
+	)
+
+	uploadContext := context.Background()
+	for _, fb := range s.fileBundlers {
+		fb.Launch(uploadContext)
+	}
```

Comment:
I don't see a good reason to have a context that is `context.Background`, it will never be done nor change, so it's rather useless. 

Can't we use the received `ctx`? Was it causing problem to you somehow?

Reply:
It wasn't causing any problems. This is the way it's used on `graph-load`, I don't know if they have any issues related to context.